doc/tunnel/SAI-Proposal-IP-in-MPLS-Tunnel.md

```diff
@@ -0,0 +1,776 @@
+}
+```
+
+
+#### fdb entries
+
+```
+fdb_a = {
+    {switch_id, vlan_1_a, mac_a},
+    {
+        {SAI_FDB_ENTRY_ATTR_TYPE, SAI_FDB_ENTRY_TYPE_STATIC},
+        {SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID, port_1_a}
+    }
+}
+fdb_2 = {
+    {switch, vlan_1_2, mac_2},
+    {
+        {SAI_FDB_ENTRY_ATTR_TYPE, SAI_FDB_ENTRY_TYPE_STATIC},
+        {SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID, port_1_2}
+    }
+}
+fdb_4 = {
+    {switch, vlan_1_4, mac_4},
+    {
+        {SAI_FDB_ENTRY_ATTR_TYPE, SAI_FDB_ENTRY_TYPE_STATIC},
+        {SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID, port_1_4}
+    }
+}
+```
+
+#### neighbor entries
+
+```
+neighbor_a = {
+    {switch_id, rif_1_a, 101.101.0.101},
+    {
+        {SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS, mac_a}
+    }
+}
+neighbor_2 = {
+    {switch_id, rif_1_2, 1.1.1.21},
+    {
+        {SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS, mac_2},
+    }
+}
+neighbor_4 = {
+    {switch_id, rif_1_4, 1.1.1.41},
+    {
+        {SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS, mac_4}
+    }
+}
+```
+
+#### decap tunnel
+
+```
+decap_tunnel = {
+    {SAI_TUNNEL_ATTR_TYPE, SAI_TUNNEL_TYPE_MPLS},
+    {SAI_TUNNEL_ATTR_UNDERLAY_INTERFACE, mpls_rif},
+    {SAI_TUNNEL_ATTR_DECAP_TTL_MODE, SAI_TUNNEL_TTL_MODE_UNIFORM_MODEL},
+    {SAI_TUNNEL_ATTR_DECAP_EXP_MODE, SAI_TUNNEL_EXP_MODE_UNIFORM_MODEL},
+}
+```
+
+#### encap tunnel
+
+```
+encap_tunnel = {
+    {SAI_TUNNEL_ATTR_TYPE, SAI_TUNNEL_TYPE_MPLS},
+    {SAI_TUNNEL_ATTR_ENCAP_TTL_MODE, SAI_TUNNEL_TTL_MODE_UNIFORM_MODEL},
+    {SAI_TUNNEL_ATTR_ENCAP_EXP_MODE, SAI_TUNNEL_EXP_MODE_UNIFORM_MODEL},
+}
+```
+
+#### next hops
+
+```
+nh_a = {
+    {SAI_NEXT_HOP_ATTR_TYPE, SAI_NEXT_HOP_TYPE_IP},
+    {SAI_NEXT_HOP_ATTR_IP, "101.101.0.101"},
+    {SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID, rif_1_a},
+}
+nh_2 = {
+    {SAI_NEXT_HOP_ATTR_TYPE, SAI_NEXT_HOP_TYPE_MPLS},
+    {SAI_NEXT_HOP_ATTR_IP, "1.1.1.21"},
+    {SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID, rif_1_2},
+    {SAI_NEXT_HOP_ATTR_LABELSTACK, [5, 3]},
+    {SAI_NEXT_HOP_ATTR_MPLS_ENCAP_TUNNEL_ID, encap_tunnel}
+}
+nh_4 = {
+    {SAI_NEXT_HOP_ATTR_TYPE, SAI_NEXT_HOP_TYPE_MPLS},
+    {SAI_NEXT_HOP_ATTR_IP, "1.1.1.41"},
+    {SAI_NEXT_HOP_ATTR_ROUTER_INTERFACE_ID, rif_1_4},
+    {SAI_NEXT_HOP_ATTR_LABELSTACK, [201]},
+    {SAI_NEXT_HOP_ATTR_MPLS_ENCAP_TUNNEL_ID, encap_tunnel}
+}
+nhg201 = {
+    {SAI_NEXT_HOP_GROUP_ATTR_NEXT_HOP_COUNT, 2},
+    {SAI_NEXT_HOP_GROUP_ATTR_NEXT_HOP_MEMBER_LIST, [nh_2, nh_4]},
+    {SAI_NEXT_HOP_GROUP_ATTR_TYPE, SAI_NEXT_HOP_GROUP_TYPE_ECMP}
+}
+```
+
+#### routes
+
+```
+route201 = {
+    {switch_id, vr0, "201.201.0.0/24"},
+    {SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID, nhg201}
+}
+route101 = {
+    {switch_id, vr0, "101.101.0.0/24"},
+    {SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID, nh_a}
+}
+```
+
+##### inseg entries
+
+```
+inseg1 = {
+   {switch_id, 1},
+   {SAI_INSEG_ENTRY_ATTR_NUM_OF_POP, 1},
+   {SAI_INSEG_ENTRY_ATTR_NEXT_HOP_ID, mpls_rif}
+}
+inseg101 = {
+   {switch_id, 101},
```

Comment:
where is this id 101 get defined in the example?

Reply:
101 is a label programmed by application.